job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java

```diff
@@ -160,19 +164,33 @@ private static void migrate(MigrateCommand command, WritePType writeType, FileSy
         CreateFilePOptions.newBuilder().setWriteType(writeType).build();
     OpenFilePOptions openFileOptions =
         OpenFilePOptions.newBuilder().setReadType(ReadPType.NO_CACHE).build();
-    try (FileInStream in = fileSystem.openFile(new AlluxioURI(source), openFileOptions);
-         FileOutStream out = fileSystem.createFile(new AlluxioURI(destination), createOptions)) {
+    final AlluxioURI destinationURI = new AlluxioURI(destination);
+    boolean retry;
+    do {
+      retry = false;
       try {
-        IOUtils.copyLarge(in, out, new byte[8 * Constants.MB]);
-      } catch (Throwable t) {
-        try {
-          out.cancel();
-        } catch (Throwable t2) {
-          t.addSuppressed(t2);
+        try (FileInStream in = fileSystem.openFile(new AlluxioURI(source), openFileOptions);
+             FileOutStream out = fileSystem.createFile(destinationURI, createOptions)) {
+          try {
+            IOUtils.copyLarge(in, out, new byte[8 * Constants.MB]);
+          } catch (Throwable t) {
+            try {
+              out.cancel();
+            } catch (Throwable t2) {
+              t.addSuppressed(t2);
+            }
+            throw t;
+          }
+        }
+      } catch (FileAlreadyExistsException e) {
+        if (overwrite) {
+          fileSystem.delete(destinationURI);
+          retry = true;
+        } else {
+          throw e;
         }
-        throw t;
       }
-    }
+    } while (retry);
```

Comment:
Should we limit the number of retries? Would it be too expensive to first check if the file exists, delete on override = true, and leave the retry handling to the job service instead of at the task level?

Reply:
In order for one to reach retry repeatedly, one must hit FileAlreadyExistsException repeatedly. The only way you can do that is for fileSystem.delete to fail which would exit the retry loop.

This isn't a real retry. This retry is only a retry when the destination file happens to already exists. The only reason why it's designed this way is because this is faster than doing a file exists check in the common case of when it's not overwriting.